src/clusterfuzz/_internal/bot/tasks/trials.py

```diff
@@ -61,17 +75,25 @@ def __init__(self):
       for config in trials_config:
         if config['app_name'] != app_name:
           continue
-        self.trials[config['app_args']] = config['probability']
+        self.trials[config['app_args']] = AppArgs(config['probability'],
+                                                  config.get('contradicts', []))
     except Exception as e:
       logs.log_warn('Unable to parse config file: %s' % str(e))
       return
 
   def setup_additional_args_for_app(self):
     """Select additional args for the specified app at random."""
-    trial_args = [
-        app_args for app_args, probability in self.trials.items()
-        if random.random() < probability
-    ]
+    trial_args = []
+    contradicts = set()
+
+    for app_args, flag_data in sorted(
+        self.trials.items(), key=lambda x: random.random()):
```

Comment:
Not sure I follow here. Do you want to do random sorting? I thought the key should be stable...

How about:
random.sample(self.trials.items(), k=len(self.trials))

Note from the two ideas I had about how to resolve the probability bias, this is probably the more costly.

Reply:
done, we shuffle the keys now